File: scripts/telegraph_publisher.ts
/**
 * Telegraph Publisher - TypeScript Implementation
 * Supports both Deno and Node.js runtimes
 */

import type { RequestInit } from "https://deno.land/std@0.208.0/http/http_status.ts";

interface TelegraphConfig {
  apiUrl: string;
  maxRetries: number;
  retryDelay: number;
  timeout: number;
  defaultAuthor: string;
  defaultAuthorUrl: string;
}

interface TelegraphResponse<T = any> {
  ok: boolean;
  result?: T;
  error?: string;
}

interface PageResult {
  url: string;
  path: string;
  title: string;
  content?: any[];
}

interface AccountResult {
  short_name: string;
  access_token: string;
  auth_url: string;
}

interface MediaResult {
  src: string;
}

/**
 * TelegraphPublisher class - Main wrapper for Telegraph API
 */
export class TelegraphPublisher {
  private config: TelegraphConfig;
  private accessToken?: string;
  private accountInfo?: AccountResult;

  constructor(accessToken?: string, config?: Partial<TelegraphConfig>) {
    this.config = {
      apiUrl: "https://api.telegra.ph",
      maxRetries: 3,
      retryDelay: 1.0,
      timeout: 30000,
      defaultAuthor: "OpenClaw",
      defaultAuthorUrl: "https://github.com/anthropics/openclaw",
      ...config,
    };

    this.accessToken = accessToken;

    console.log(
      `[INFO] Initializing Telegraph Publisher (API: ${this.config.apiUrl})`
    );
  }

  /**
   * Make HTTP request to Telegraph API with exponential backoff
   */
  private async request<T = any>(
    method: string,
    endpoint: string,
    data?: Record<string, any>,
    retries: number = 0
  ): Promise<T> {
    const url = `${this.config.apiUrl}${endpoint}`;
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      "User-Agent": "OpenClaw/1.0",
    };

    try {
      const requestOptions: RequestInit = {
        method,
        headers,
      };

      if (method === "POST" && data) {
        requestOptions.body = JSON.stringify(data);
      }

      // Use native fetch (works in both Deno and Node.js 18+)
      const response = await fetch(url, requestOptions);

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }

      const result: TelegraphResponse<T> = await response.json();

      if (!result.ok) {
        throw new Error(`Telegraph API error: ${result.error || "Unknown error"}`);
      }

      return result.result as T;
    } catch (error) {
      if (retries < this.config.maxRetries) {
        const waitTime =
          this.config.retryDelay * Math.pow(2, retries); // Exponential backoff
        console.warn(
          `[WARN] Request failed (${retries + 1}/${this.config.maxRetries}): ${error}. Retrying in ${waitTime}s...`
        );
        await this.sleep(waitTime * 1000);
        return this.request<T>(method, endpoint, data, retries + 1);
      } else {
        console.error(
          `[ERROR] Request failed after ${this.config.maxRetries} retries: ${error}`
        );
        throw error;
      }
    }
  }

  /**
   * Sleep for milliseconds (works in both Deno and Node.js)
   */
  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  /**
   * Create Telegraph account and return access token
   */
  async createAccount(
    shortName?: string,
    authorName?: string,
    authorUrl?: string
  ): Promise<string> {
    if (!shortName) {
      // Generate unique short name
      const timestamp = Math.floor(Date.now() / 1000);
      shortName = `openclaw_${timestamp}`;
    }

    const data = {
      short_name: shortName,
      author_name: authorName || this.config.defaultAuthor,
      author_url: authorUrl || this.config.defaultAuthorUrl,
    };

    console.log(
      `[INFO] Creating Telegraph account: ${shortName}`
    );
    const result = await this.request<AccountResult>(
      "POST",
      "/createAccount",
      data
    );

    this.accessToken = result.access_token;
    this.accountInfo = result;

    console.log(
      `[INFO] Account created successfully. Token: ${this.accessToken.substring(0, 20)}...`
    );
    return this.accessToken;
  }

  /**
   * Publish Markdown content to Telegraph
   */
  async publishMarkdown(
    title: string,
    content: string,
    author?: string,
    authorUrl?: string
  ): Promise<string> {
    if (!this.accessToken) {
      await this.createAccount();
    }

    // Convert markdown to HTML
    const htmlContent = this.markdownToHtml(content);

    return this.publishHtml(title, htmlContent, author, authorUrl);
  }

  /**
   * Publish HTML content to Telegraph
   */
  async publishHtml(
    title: string,
    content: string | Record<string, any>[],
    author?: string,
    authorUrl?: string
  ): Promise<string> {
    if (!this.accessToken) {
      await this.createAccount();
    }

    // Parse content into Telegraph format if it's HTML string
    let parsedContent: Record<string, any>[] = Array.isArray(content)
      ? content
      : [{ tag: "p", children: [content] }];

    const data: Record<string, any> = {
      access_token: this.accessToken,
      title: title.substring(0, 256), // Telegraph limit
      author_name: author || this.config.defaultAuthor,
      author_url: authorUrl || this.config.defaultAuthorUrl,
      content: parsedContent,
      return_content: true,
    };

    console.log(`[INFO] Publishing page: ${title}`);
    const result = await this.request<PageResult>(
      "POST",
      "/createPage",
      data
    );

    const url = result.url;
    console.log(`[INFO] Page published successfully: ${url}`);
    return url;
  }

  /**
   * Update existing Telegraph page
   */
  async updatePage(
    path: string,
    content: string | Record<string, any>[],
    title?: string,
    author?: string,
    authorUrl?: string
  ): Promise<string> {
    if (!this.accessToken) {
      throw new Error("Access token required for updating pages");
    }

    // Parse content
    let parsedContent: Record<string, any>[] = Array.isArray(content)
      ? content
      : [{ tag: "p", children: [content] }];

    const data: Record<string, any> = {
      access_token: this.accessToken,
      path,
      content: parsedContent,
      return_content: true,
    };

    if (title) data.title = title.substring(0, 256);
    if (author) data.author_name = author;
    if (authorUrl) data.author_url = authorUrl;

    console.log(`[INFO] Updating page: ${path}`);
    const result = await this.request<PageResult>(
      "POST",
      "/editPage",
      data
    );

    const url = result.url;
    console.log(`[INFO] Page updated successfully: ${url}`);
    return url;
  }

  /**
   * Retrieve Telegraph page
   */
  async getPage(
    path: string,
    returnContent: boolean = true
  ): Promise<PageResult> {
    const data = {
      path,
      return_content: returnContent,
    };

    console.log(`[INFO] Fetching page: ${path}`);
    return this.request<PageResult>("GET", `/getPage/${path}`, data);
  }

  /**
   * Upload image or media file to Telegraph
   */
  async uploadMedia(filePath: string): Promise<string> {
    console.log(`[INFO] Uploading media: ${filePath}`);

    // Read file (works in both Deno and Node.js with conditional imports)
    let fileBuffer: Uint8Array;

    try {
      // Try Deno API first
      if (typeof Deno !== "undefined") {
        fileBuffer = await Deno.readFile(filePath);
      } else {
        // Node.js fallback
        const fs = await import("fs");
        const fsPromises = fs.promises;
        fileBuffer = await fsPromises.readFile(filePath);
      }
    } catch (error) {
      throw new Error(`Failed to read file: ${filePath}`);
    }

    // Create FormData and upload
    const formData = new FormData();
    const blob = new Blob([fileBuffer]);
    formData.append("file", blob, filePath);

    try {
      const response = await fetch(`${this.config.apiUrl}/upload`, {
        method: "POST",
        body: formData,
      });

      const result: MediaResult[] = await response.json();

      if (!Array.isArray(result) || result.length === 0) {
        throw new Error(`Media upload failed: ${JSON.stringify(result)}`);
      }

      const mediaUrl = `https://telegra.ph${result[0].src}`;
      console.log(`[INFO] Media uploaded: ${mediaUrl}`);
      return mediaUrl;
    } catch (error) {
      throw new Error(`Media upload error: ${error}`);
    }
  }

  /**
   * Convert basic Markdown to HTML
   */
  private markdownToHtml(markdown: string): string {
    let html = markdown;

    // Headers
    html = html.replace(/^### (.*?)$/gm, "<h3>$1</h3>");
    html = html.replace(/^## (.*?)$/gm, "<h2>$1</h2>");
    html = html.replace(/^# (.*?)$/gm, "<h1>$1</h1>");

    // Bold and italic
    html = html.replace(/\*\*(.*?)\*\*/g, "<strong>$1</strong>");
    html = html.replace(/\*(.*?)\*/g, "<em>$1</em>");
    html = html.replace(/__(.*?)__/g, "<strong>$1</strong>");
    html = html.replace(/_(.*?)_/g, "<em>$1</em>");

    // Code blocks
    html = html.replace(/```(.*?)```/gs, "<pre><code>$1</code></pre>");
    html = html.replace(/`([^`]*?)`/g, "<code>$1</code>");

    // Links
    html = html.replace(/\[(.*?)\]\((.*?)\)/g, '<a href="$2">$1</a>');

    // Line breaks
    html = html.replace(/\n\n/g, "</p><p>");
    html = `<p>${html}</p>`;

    // Lists
    html = html.replace(/^\* (.*?)$/gm, "<li>$1</li>");

    return html;
  }

  /**
   * Save access token to file
   */
  async saveToken(filePath: string): Promise<void> {
    if (!this.accessToken) {
      throw new Error("No access token to save");
    }

    const data = {
      access_token: this.accessToken,
      account_info: this.accountInfo,
      created_at: new Date().toISOString(),
    };

    try {
      // Try Deno API first
      if (typeof Deno !== "undefined") {
        const fileUrl = new URL("file:///" + filePath).href;
        await Deno.writeTextFile(filePath, JSON.stringify(data, null, 2));
      } else {
        // Node.js fallback
        const fs = await import("fs");
        const fsPromises = fs.promises;
        await fsPromises.writeFile(filePath, JSON.stringify(data, null, 2));
      }
      console.log(`[INFO] Token saved to: ${filePath}`);
    } catch (error) {
      throw new Error(`Failed to save token: ${error}`);
    }
  }

  /**
   * Load access token from file
   */
  static async loadToken(filePath: string): Promise<string> {
    try {
      let content: string;

      // Try Deno API first
      if (typeof Deno !== "undefined") {
        content = await Deno.readTextFile(filePath);
      } else {
        // Node.js fallback
        const fs = await import("fs");
        const fsPromises = fs.promises;
        content = await fsPromises.readFile(filePath, "utf-8");
      }

      const data = JSON.parse(content);
      return data.access_token;
    } catch (error) {
      throw new Error(`Failed to load token: ${error}`);
    }
  }
}

export default TelegraphPublisher;

File: scripts/telegraph_publisher.py
#!/usr/bin/env python3
"""
Telegraph Publisher - Publish formatted content to Telegraph.ph
Used for OpenClaw subagent output, HEARTBEAT reports, and manual publishing.
"""

import requests
import json
import os
import time
import re
from pathlib import Path
from typing import Optional, Dict, Any
from datetime import datetime
import logging

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s [%(levelname)s] %(message)s',
    handlers=[
        logging.FileHandler(os.path.expanduser("~/.openclaw/logs/telegraph.log")),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger(__name__)

class TelegraphPublisher:
    """Publish content to Telegraph with retry logic and error handling."""
    
    def __init__(self, config_path: Optional[str] = None):
        """Initialize publisher with config."""
        self.config_path = config_path or os.path.expanduser("~/.openclaw/workspace/config/telegraph.json")
        self.config = self._load_config()
        self.token = self._load_token()
        self.api_endpoint = self.config['api']['endpoint']
        self.timeout = self.config['api']['timeout_seconds']
        
    def _load_config(self) -> Dict[str, Any]:
        """Load Telegraph configuration."""
        if not os.path.exists(self.config_path):
            raise FileNotFoundError(f"Config not found: {self.config_path}")
        
        with open(self.config_path, 'r') as f:
            return json.load(f)
    
    def _load_token(self) -> str:
        """Load access token from secure storage."""
        token_path = os.path.expanduser(self.config['token']['storage'])
        
        if not os.path.exists(token_path):
            raise FileNotFoundError(f"Telegraph token not found: {token_path}")
        
        with open(token_path, 'r') as f:
            return f.read().strip()
    
    def _retry_request(self, method: str, endpoint: str, **kwargs) -> Dict[str, Any]:
        """Make request with exponential backoff retry."""
        retry_config = self.config['api']['retry']
        max_attempts = retry_config['max_attempts']
        initial_delay = retry_config['initial_delay_ms'] / 1000
        max_delay = retry_config['max_delay_ms'] / 1000
        
        delay = initial_delay
        last_error = None
        
        for attempt in range(max_attempts):
            try:
                if method.upper() == 'POST':
                    response = requests.post(endpoint, timeout=self.timeout, **kwargs)
                else:
                    response = requests.get(endpoint, timeout=self.timeout, **kwargs)
                
                if response.status_code == 200:
                    return response.json()
                else:
                    last_error = f"HTTP {response.status_code}: {response.text}"
                    
                    if response.status_code == 429:  # Rate limit
                        delay = min(delay * 2, max_delay)
                        logger.warning(f"Rate limited, retry in {delay:.1f}s")
                        time.sleep(delay)
                        continue
                    
                    raise Exception(last_error)
            
            except requests.Timeout:
                last_error = f"Timeout after {self.timeout}s"
                if attempt < max_attempts - 1:
                    time.sleep(delay)
                    delay = min(delay * 2, max_delay)
            
            except Exception as e:
                last_error = str(e)
                if attempt < max_attempts - 1:
                    time.sleep(delay)
                    delay = min(delay * 2, max_delay)
        
        raise Exception(f"Failed after {max_attempts} attempts: {last_error}")
    
    def _markdown_to_telegraph_content(self, markdown: str) -> list:
        """Convert Markdown to Telegraph content format."""
        content = []
        lines = markdown.split('\n')
        
        i = 0
        while i < len(lines):
            line = lines[i]
            
            # Headings (H2-H6)
            if line.startswith('## '):
                content.append({'tag': 'h3', 'children': [line[3:].strip()]})
            elif line.startswith('### '):
                content.append({'tag': 'h4', 'children': [line[4:].strip()]})
            elif line.startswith('#### '):
                content.append({'tag': 'h5', 'children': [line[5:].strip()]})
            
            # Code blocks
            elif line.startswith('```'):
                code_lines = []
                i += 1
                while i < len(lines) and not lines[i].startswith('```'):
                    code_lines.append(lines[i])
                    i += 1
                
                code_content = '\n'.join(code_lines).strip()
                if code_content:
                    content.append({'tag': 'pre', 'children': [code_content]})
            
            # Tables (basic support)
            elif '|' in line and i + 1 < len(lines) and '|' in lines[i + 1]:
                # Simple table to paragraph conversion for now
                content.append({'tag': 'p', 'children': [line]})
            
            # Lists
            elif line.startswith('- '):
                content.append({'tag': 'p', 'children': ['• ' + line[2:].strip()]})
            elif line.startswith('* '):
                content.append({'tag': 'p', 'children': ['• ' + line[2:].strip()]})
            
            # Paragraphs
            elif line.strip():
                # Inline formatting: bold (**text**) and italic (*text*)
                formatted = line.strip()
                formatted = re.sub(r'\*\*(.+?)\*\*', r'<strong>\1</strong>', formatted)
                formatted = re.sub(r'\*(.+?)\*', r'<em>\1</em>', formatted)
                content.append({'tag': 'p', 'children': [formatted]})
            
            i += 1
        
        return content if content else [{'tag': 'p', 'children': ['(empty content)']}]
    
    def publish_markdown(self, title: str, markdown_content: str, 
                        return_content: bool = False) -> Dict[str, Any]:
        """
        Publish Markdown content to Telegraph.
        
        Args:
            title: Article title
            markdown_content: Markdown formatted content
            return_content: If True, return full Telegraph content
        
        Returns:
            Dict with 'url' key and Telegraph response
        """
        logger.info(f"Publishing: {title}")
        
        if not markdown_content.strip():
            raise ValueError("Content cannot be empty")
        
        # Convert to Telegraph content format
        content = self._markdown_to_telegraph_content(markdown_content)
        
        # Create page
        url = f"{self.api_endpoint}/createPage"
        payload = {
            'access_token': self.token,
            'title': title[:256],  # Telegraph limit
            'author_name': self.config['author']['name'],
            'author_url': 'https://github.com/rreilly/openclaw',
            'content': content,
            'return_content': return_content
        }
        
        try:
            response = self._retry_request('POST', url, json=payload)
            
            if 'result' in response:
                result = response['result']
                page_url = result['url']
                logger.info(f"✅ Published: {page_url}")
                
                return {
                    'success': True,
                    'url': f"https://telegra.ph{page_url}",
                    'path': result.get('path'),
                    'title': result.get('title')
                }
            else:
                error_msg = response.get('error', 'Unknown error')
                logger.error(f"❌ API Error: {error_msg}")
                return {'success': False, 'error': error_msg}
        
        except Exception as e:
            logger.error(f"❌ Publish failed: {str(e)}")
            return {'success': False, 'error': str(e)}
    
    def publish_html(self, title: str, html_content: str,
                    return_content: bool = False) -> Dict[str, Any]:
        """Publish HTML content to Telegraph."""
        logger.info(f"Publishing HTML: {title}")
        
        url = f"{self.api_endpoint}/createPage"
        payload = {
            'access_token': self.token,
            'title': title[:256],
            'author_name': self.config['author']['name'],
            'author_url': 'https://github.com/rreilly/openclaw',
            'content': html_content,
            'return_content': return_content
        }
        
        try:
            response = self._retry_request('POST', url, json=payload)
            
            if 'result' in response:
                result = response['result']
                page_url = result['url']
                logger.info(f"✅ Published: {page_url}")
                
                return {
                    'success': True,
                    'url': f"https://telegra.ph{page_url}",
                    'path': result.get('path'),
                    'title': result.get('title')
                }
            else:
                error_msg = response.get('error', 'Unknown error')
                logger.error(f"❌ API Error: {error_msg}")
                return {'success': False, 'error': error_msg}
        
        except Exception as e:
            logger.error(f"❌ Publish failed: {str(e)}")
            return {'success': False, 'error': str(e)}
    
    def test_connectivity(self) -> bool:
        """Test Telegraph API connectivity."""
        logger.info("Testing Telegraph API connectivity...")
        
        try:
            url = f"{self.api_endpoint}/getPageViews"
            response = requests.get(url, params={'path': 'test'}, timeout=self.timeout)
            
            if response.status_code == 200:
                logger.info("✅ Telegraph API is accessible")
                return True
            else:
                logger.warning(f"⚠️  Telegraph API returned {response.status_code}")
                return False
        
        except Exception as e:
            logger.error(f"❌ Cannot reach Telegraph API: {str(e)}")
            return False


def main():
    """CLI entry point for testing."""
    import sys
    
    if len(sys.argv) < 2:
        print("Usage: telegraph_publisher.py <command> [args]")
        print("\nCommands:")
        print("  test              - Test API connectivity")
        print("  publish-md <title> <file>  - Publish Markdown file")
        print("  publish-html <title> <file> - Publish HTML file")
        sys.exit(1)
    
    command = sys.argv[1]
    
    try:
        pub = TelegraphPublisher()
        
        if command == 'test':
            result = pub.test_connectivity()
            sys.exit(0 if result else 1)
        
        elif command == 'publish-md':
            if len(sys.argv) < 4:
                print("Usage: telegraph_publisher.py publish-md <title> <file>")
                sys.exit(1)
            
            title = sys.argv[2]
            filepath = sys.argv[3]
            
            with open(filepath, 'r') as f:
                content = f.read()
            
            result = pub.publish_markdown(title, content)
            
            if result['success']:
                print(f"\n✅ Published: {result['url']}")
            else:
                print(f"\n❌ Error: {result['error']}")
                sys.exit(1)
        
        else:
            print(f"Unknown command: {command}")
            sys.exit(1)
    
    except Exception as e:
        print(f"❌ Error: {str(e)}")
        sys.exit(1)


if __name__ == '__main__':
    main()
